Let PacketWriter::write_packet use buffers convertible to Cow

AlexTMjugador opened the pull request:
As pointed out in issue #19, in some cases it may not be desirable to heap allocate Ogg packet contents, but the current parameters of the `write_packet` method impose a heap allocation no matter what.

However, the `PacketWriter` struct doesn't actually need to own the packet data, as long as it stays in memory unmodified during its lifetime. The `Cow` smart pointer that the standard library provides captures this requirement more precisely, and provides for more elegant, flexible and potentially efficient client code.

Therefore, substitute `Box` for `Cow` in the aforementioned code. Boxed slices can still be used with this new design, although with a different syntax, which breaks backwards compatibility.

AlexTMjugador committed:
Let PacketWriter::write_packet use buffers convertible to Cow

As pointed out in issue #19, in some cases it may not be desirable to
heap allocate Ogg packet contents, but the current parameters of the
`write_packet` method impose a heap allocation no matter what.

However, the `PacketWriter` struct doesn't actually need to own the
packet data, as long as it stays in memory unmodified during its
lifetime. The `Cow` smart pointer that the standard library provides
captures this requirement more precisely, and provides for more elegant,
flexible and potentially efficient client code.

Therefore, substitute `Box` for `Cow` in the aforementioned code. Boxed
slices can still be used with this new design, although with a different
syntax, which breaks backwards compatibility.

est31 commented on src/writing.rs:
```diff
@@ -118,15 +119,16 @@ impl <T :io::Write> PacketWriter<T> {
-			absgp :u64) -> IoResult<()> {
+	pub fn write_packet<P: Into<Cow<'writer, [u8]>>>(&mut self, pck_cont :P,
+		serial :u32,
+		inf :PacketWriteEndInfo,
```
Can you leave this indented two times?

AlexTMjugador replied:
Okay, I'll do it!

est31 reviewed (changes requested):
Thanks for the PR! I wonder if it's possible to have something more general than `Cow`. Maybe `AsRef<[u8]>` or something?

AlexTMjugador commented:
> I wonder if it's possible to have something more general than `Cow`. Maybe `AsRef<[u8]>` or something?

This sounds like a good idea, thanks! However, after swapping the generic `Into<Cow<[u8]>>` parameter for `AsRef<[u8]>`, I found the result less ergonomic, because the `write_packet` method would need to always take a reference to a byte buffer, and not be able to accept an owned buffer like `Vec<u8>` (otherwise Rust would drop the maybe owned `AsRef<[u8]>` parameter at the end of the method body, which does not compile).

On the other hand, with the current approach, the caller can just convert whatever type that implements `AsRef<[u8]>` beforehand to a byte slice by calling `as_ref` on that type, and things would still work fine because a `Cow` can then be constructed from that borrowed byte slice. So I'd argue that it's still flexible enough to accept a wide variety of byte buffers, although as I explained before the lifetime requirements are a bit longer than a user may expect, so you'd need to call `as_ref` relatively "early".

Another advantage I can see from using a `Cow` is that, in the very unlikely case that this library is changed so there is a need to modify packet data before writing it, a `Cow` can also support that as efficiently as possible, because if the user already provided a `Vec<u8>` to the method, its memory buffer will be reused without any additional cost.

AlexTMjugador committed:
Change identation as per review

AlexTMjugador committed:
Fix CI build error by not using an anonymous lifetime

Newer Rust versions work introduced additional lifetime inference rules
that made this code work, but they are not available in the MSRV this
library targets.

est31 commented:
Good point about the need for a generic param. The PR should be good now but I won't merge it for now because it is a breaking change of the API. I want to make a new release of the ogg crate once the 2021 edition lands though.

AlexTMjugador commented:
That's understandable, thanks. However, right now it's bugging me a bit that the CI build for Rust 1.27.2 does not pass, as I've developed this PR using the latest Rust version and I used some syntax that simply didn't compile back then. Do you plan on bumping MSRV on the next release? Should I continue to fix compilation with this older Rust version?

AlexTMjugador committed:
Fix compilation with async feature

est31 commented:
@AlexTMjugador no, the MSRV will be increased to 1.56.0.

AlexTMjugador committed:
Revert "Fix CI build error by not using an anonymous lifetime"

This reverts commit 01ee00e04cf81abccb9f969a038cdafc4322b522.

Fixing this build error is no longer necessary because MSRV will be
bumped to the current stable version, which accepts this shorter syntax.
